### DISCUSSION TITLE
What is benefit of dual buffer

### DISCUSSION THREAD

**Discussion by AlterSb** [mrcodetastic/ESP32-HUB75-MatrixPanel-DMA#490] [General]
Hi, I got a question. What is benefit of 'mxconfig.double_buff ' ? Why some people use it? There is no any documentation on this one.

**Comment by mrcodetastic** [comment]
You can do complex drawing routines in the 'background' and then 'flip' DMA buffers.

Most people don't need it. People that need it, know why they need it.

**Reply by AlterSb** [reply]
Thank you, as you said 'Most people don't need it' and yes, normal routine work perfectly fine for me but would you kind tell me an example and why normal routine doesn't work well on that example. I want to learn from this. All I can think of is complex drawing that doesn't need a continous display update? correct me pls.

**Reply by mrcodetastic** [reply]
From chatgpt:

Double buffering is a technique used in computer graphics to eliminate flickering and provide smoother visuals when rendering images or animations on a display. It involves having two separate buffers or memory areas: one is the "front buffer" that the user sees, and the other is the "back buffer" where new images or frames are drawn.

The process works like this:
1. The application draws or updates the image/frame in the back buffer.
2. Once the new frame is fully prepared, the front and back buffers are swapped.
3. The previously back buffer becomes the new front buffer and is displayed, while the old front buffer becomes the new back buffer for the next frame.

This swapping happens quickly, so users perceive the transition as smooth. Double buffering prevents flickering because the display is never updated while a new frame is partially drawn, reducing the likelihood of displaying an incomplete or partially-rendered frame.

Overall, double buffering enhances the visual quality and user experience by reducing visual artifacts and providing smoother animations in computer graphics applications.

**Reply by AlterSb** [reply]
Thanks you, I have tried to implement that in my existed code. Seem to be a lot of flickering. It's probably because my code wasn't for double buffer from beginning. But thank you, I will learn more of this.

**Reply by daveythacher** [reply]
It should be done at the top of the frame or the vsync point.

**Comment by daveythacher** [comment]
It done to prevent tearing. It allows the frame to change completely at one time. Otherwise there are two frames on the screen at the same time.

**Comment by ukdepartureboards** [comment]
Create a simple sketch that just scrolls some text horizontally, you can see night and day the difference between enabling and not enabling double buffering.
